ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/ShelleyHFC.hs

```diff
@@ -137,9 +137,8 @@ shelleyTransition ::
 shelleyTransition ShelleyPartialLedgerConfig{..}
                   transitionMajorVersionRaw
                   state =
-      takeAny
-    . mapMaybe isTransition
-    . Shelley.Inspect.protocolUpdates genesis
+      isTransition
+    . Shelley.Inspect.pparamsUpdate
```

Comment:
`pparamsUpdate` calls `futurePParamsGovStateG`, which according to the documentation in https://github.com/IntersectMBO/cardano-ledger/pull/4253 has the following behavior:

> This getter is only reliable and efficient 2 stability before the end of the epoch. Depending on the era, if called earlier in the epoch, it will either produce unreliable results or getting those results will be somewhat costly.

`shelleyTransition` will be invoked on every forecast/tick (via `singleEraTransition`), so it will be called in particular in the first 4k/f of an epoch. If that is undesirable, it would be fine for us if `futurePParamsGovStateG` would just always immediately return `Nothing` in that case.

Reply:
I think to properly solve this is to run
```haskell
  guard $ shelleyAfterVoting >= fromIntegral k
```
before we look into `pparamsUpdate`, right?

Reply:
>  it would be fine for us if futurePParamsGovStateG would just always immediately return Nothing in that case.

This is quite a bit harder to do on the ledger side, unfortunately. I was hoping it is possible to not force this calculation before it is really needed

Reply:
Giving it a bit more though, I think I know a nice way to solve this in ledger. Thank you for the feedback

Reply:
New and improved approach has been implemented in https://github.com/IntersectMBO/cardano-ledger/pull/4253

With this new approach we no longer need to worry about any of the conditions that we previously had to worry about in here:
* quorum of same votes on proposals have been reached
* are we in the correct part of the epoch (two stability windows before the end of the epoch or not)

This is all now taken care of by the ledger with the help of `nextEpochUpdatedPParams` that is cheap and safe to be called through any point of the epoch.